diffusion/models/autoencoder.py

```diff
@@ -0,0 +1,613 @@
+        disc_weight *= self.discriminator_weight
+        # Set the discriminator weight. It should be negative to reverse gradients into the autoencoder.
+        self.scale_gradients.set_scale(-disc_weight.item())
+        return disc_weight, nll_grads_norm, disc_grads_norm
+
+    def forward(self, outputs: Dict[str, torch.Tensor], batch: Dict[str, torch.Tensor],
+                last_layer: torch.Tensor) -> Dict[str, torch.Tensor]:
+        losses = {}
+        # Basic L1 reconstruction loss
+        ae_loss = F.l1_loss(outputs['x_recon'], batch[self.input_key], reduction='none')
+        # Count the number of output elements to normalize the loss
+        num_output_elements = ae_loss.numel() // ae_loss.shape[0]
+        losses['ae_loss'] = ae_loss.mean()
+
+        # LPIPs loss. Images for LPIPS must be in [-1, 1]
+        recon_img = outputs['x_recon'].clamp(-1, 1)
+        target_img = batch[self.input_key].clamp(-1, 1)
+        lpips_loss = self.lpips(recon_img, target_img)
+        losses['lpips_loss'] = lpips_loss.mean()
+
+        # Make the nll loss
+        rec_loss = ae_loss + self.lpips_weight * lpips_loss
+        # Note: the +2 here comes from the nll of the laplace distribution.
+        # It's only here to make you feel better by keeping the loss positive for longer.
+        nll_loss = rec_loss / torch.exp(self.log_var) + self.log_var + 2
+        nll_loss = nll_loss.mean()
+        losses['nll_loss'] = nll_loss
+        losses['output_variance'] = torch.exp(self.log_var)
+
+        # Discriminator loss
+        real = self.discriminator(batch[self.input_key])
+        fake = self.discriminator(self.scale_gradients(outputs['x_recon']))
+        real_loss = F.binary_cross_entropy_with_logits(real, torch.ones_like(real))
+        fake_loss = F.binary_cross_entropy_with_logits(fake, torch.zeros_like(fake))
+        losses['disc_real_loss'] = real_loss
+        losses['disc_fake_loss'] = fake_loss
+        losses['disc_loss'] = 0.5 * (real_loss + fake_loss)
+
+        # Update the adaptive discriminator weight
+        disc_weight, nll_grads_norm, disc_grads_norm = self.calc_discriminator_adaptive_weight(
+            nll_loss, fake_loss, last_layer)
+        losses['disc_weight'] = disc_weight
+        losses['nll_grads_norm'] = nll_grads_norm
+        losses['disc_grads_norm'] = disc_grads_norm
+
+        # Make the KL divergence loss (effectively regularize the latents)
+        mean = outputs['mean']
+        log_var = outputs['log_var']
+        kl_div_loss = -0.5 * (1 + log_var - mean.pow(2) - log_var.exp())
+        # Count the number of latent elements to normalize the loss
+        num_latent_elements = mean.numel() // kl_div_loss.shape[0]
+        losses['kl_div_loss'] = kl_div_loss.mean()
+
+        # Combine the losses. Downweight the kl_div_loss to account for differing dimensionalities.
+        dimensionality_weight = num_latent_elements / num_output_elements
```

Comment:
This is to allow using a mean instead of a sum across pixels/latent dims to make the loss scale invariant to image resolution without changing the relative loss weights. It's a choice, I can roll it back if you like.